From d9fc4a372074b1079c193c422fc4a180e79b6636 Mon Sep 17 00:00:00 2001
From: Yuxin Hu <yuxinhu@google.com>
Date: Tue, 12 Aug 2025 12:17:36 -0700
Subject: [PATCH] [M139] Fix sort uniform bug

Current sort doesn't take care of the case when both uniforms
are struct specfiers. Given that struct types are easily broken by
sort, make the sort not reordering uniforms if both are structs.

Bug: b/437825940
Change-Id: Idda1810ac4234f7e1547735e4e09658ab0a57eed
Reviewed-on: https://chromium-review.googlesource.com/c/angle/angle/+/6842936
Reviewed-by: Charlie Lao <cclao@google.com>
Reviewed-by: Amirali Abdolrashidi <abdolrashidi@google.com>
Reviewed-by: Shahbaz Youssefi <syoussefi@chromium.org>
(cherry picked from commit edb056272fd29a776f599348658ce356a67612fb)
Reviewed-on: https://chromium-review.googlesource.com/c/angle/angle/+/6860279
---

diff --git a/src/compiler/translator/Compiler.cpp b/src/compiler/translator/Compiler.cpp
index a640705..dfaf27f 100644
--- a/third_party/angle/src/compiler/translator/Compiler.cpp
+++ b/third_party/angle/src/compiler/translator/Compiler.cpp
@@ -237,12 +237,16 @@
                                       ->getAsSymbolNode()
                                       ->variable()
                                       .getType();
+        // If both uniforms are structs, do not reorder them
+        if (firstType.getStruct() != nullptr && secondType.getStruct() != nullptr)
+        {
+            return false;
+        }
         // First, sort by precision: lowp and mediump are smaller than highp
         if (firstType.getPrecision() != secondType.getPrecision())
         {
             return firstType.getPrecision() != TPrecision::EbpHigh;
         }
-
         // We don't sort highp uniforms. If both uniforms are highp, consider them as equivalent
         if (firstType.getPrecision() == TPrecision::EbpHigh &&
             secondType.getPrecision() == TPrecision::EbpHigh)
@@ -262,11 +266,6 @@
         {
             return firstType.getStruct() == nullptr;
         }
-        // If both are struct, place the one that has specifier in the front
-        if (firstType.getStruct() != nullptr && secondType.getStruct() != nullptr)
-        {
-            return firstType.isStructSpecifier();
-        }
         // criteria 3, non-matrix is smaller than matrix
         if (firstType.isMatrix() != secondType.isMatrix())
         {
diff --git a/src/tests/gl_tests/UniformTest.cpp b/src/tests/gl_tests/UniformTest.cpp
index 1615b13..d55232d 100644
--- a/third_party/angle/src/tests/gl_tests/UniformTest.cpp
+++ b/third_party/angle/src/tests/gl_tests/UniformTest.cpp
@@ -2342,6 +2342,22 @@
     glDeleteProgram(program);
 }
 
+// That that TCompiler::sortUniforms() does not break the shader code when there are multiple
+// uniforms of the struct data type, and both of them are struct specifiers, and one struct
+// references the other struct.
+TEST_P(UniformTestES31, UniformReorderDoesNotBreakStructUniformsV2)
+{
+    constexpr char kFS[] =
+        "#version 310 es\n"
+        "precision mediump float;\n"
+        "uniform struct S1 { samplerCube ar; } a1;\n"
+        "uniform struct S2 { S1 s; } a2;\n"
+        "void main (void)\n"
+        "{}";
+    GLuint program = CompileProgram(essl31_shaders::vs::Simple(), kFS);
+    ASSERT_NE(program, 0u);
+}
+
 // Test a uniform struct containing a non-square matrix and a boolean.
 // Minimal test case for a bug revealed by dEQP tests.
 TEST_P(UniformTestES3, StructWithNonSquareMatrixAndBool)
